src/audio/kpb.c

```diff
@@ -893,6 +927,165 @@ static int kpb_reset(struct comp_dev *dev)
+{
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+	uint16_t ch;
+	size_t i;
+
+	AE_SETCBEGIN0(ostream->addr);
+	AE_SETCEND0(ostream->end_addr);
+
+	buffer_stream_invalidate(source, size);
+	const ae_int16 *in_ptr = (const ae_int16 *)istream->r_ptr;
+	ae_int16x4 d16 = AE_ZERO16();
+	const size_t in_offset = in_channels * sizeof(ae_int16);
+	const size_t out_offset = micsel_channels * sizeof(ae_int16);
+	const size_t samples_per_chan = size / (sizeof(uint16_t) * micsel_channels);
+	ae_int16 *out_ptr;
+
+	for (ch = 0; ch < micsel_channels; ch++) {
+		const ae_int16 *input_data = (const ae_int16 *)(in_ptr) + offsets[ch];
+
+		out_ptr = (ae_int16 *)ostream->w_ptr;
+		out_ptr += ch;
+		for (i = 0; i < samples_per_chan; i++) {
+			AE_L16_XP(d16, input_data, in_offset);
+			AE_S16_0_XC(d16, out_ptr, out_offset);
+		}
+	}
+}
+
+static void kpb_micselect_copy32(struct comp_buffer __sparse_cache *sink,
+				 struct comp_buffer __sparse_cache *source, size_t size,
+				 uint32_t in_channels, uint32_t micsel_channels, uint32_t *offsets)
+{
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+	uint16_t ch;
+	size_t i;
+
+	AE_SETCBEGIN0(ostream->addr);
+	AE_SETCEND0(ostream->end_addr);
+
+	buffer_stream_invalidate(source, size);
+
+	const ae_int32 *in_ptr = (const ae_int32 *)istream->r_ptr;
+	ae_int32x2 d32 = AE_ZERO32();
+	const size_t in_offset = in_channels * sizeof(ae_int32);
+	const size_t out_offset = micsel_channels * sizeof(ae_int32);
+	const size_t samples_per_chan = size / (sizeof(uint32_t) * micsel_channels);
+	ae_int32 *out_ptr;
+
+	for (ch = 0; ch < micsel_channels; ch++) {
+		const ae_int32 *input_data = (const ae_int32 *)(in_ptr) + offsets[ch];
+
+		out_ptr = (ae_int32 *)ostream->w_ptr;
+		out_ptr += ch;
+		for (i = 0; i < samples_per_chan; i++) {
+			AE_L32_XP(d32, input_data, in_offset);
+			AE_S32_L_XC(d32, out_ptr, out_offset);
+		}
+	}
+}
+#else
+static void kpb_micselect_copy16(struct comp_buffer __sparse_cache *sink,
+				 struct comp_buffer __sparse_cache *source, size_t size,
+				 uint32_t in_channels,  uint32_t micsel_channels, uint32_t *offsets)
+{
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+
+	buffer_stream_invalidate(source, size);
+	size_t out_samples;
+	uint16_t ch;
+
+	const uint16_t *in_data;
+	uint16_t *out_data;
+	const uint32_t samples_per_chan = size / (sizeof(uint16_t) * micsel_channels);
+
+	for (ch = 0; ch < micsel_channels; ch++) {
+		out_samples = 0;
+		in_data = (uint16_t *)istream->r_ptr;
+		out_data = (uint16_t *)ostream->w_ptr;
+
+		for (size_t i = 0; i < samples_per_chan * in_channels; i += in_channels) {
+			if (&out_data[out_samples + ch]
+					>= (uint16_t *)ostream->end_addr) {
+				out_data = (uint16_t *)ostream->addr;
```

Comment:
actually, our the type of our signal is signed, it's better to set out_data &  in_data  as int16_t.